sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -942,9 +946,21 @@ Scheduler::GraphBuilder::addCG(std::unique_ptr<detail::CG> CommandGroup,
           NeedMemMoveToHost = true;
           MemMoveTargetQueue = HT.MQueue;
         }
-      } else if (!Queue->is_host() && !Record->MCurContext->is_host())
-        NeedMemMoveToHost = true;
-
+      } else if (!Queue->is_host() && !Record->MCurContext->is_host()) {
+        bool p2p = false;
+        Queue->getPlugin().call<PiApiKind::piextP2P>(
+            Queue->getDeviceImplPtr()->getHandleRef(),
+            findAllocaCmd(Record)
+                ->getQueue()
+                ->getDeviceImplPtr()
+                ->getHandleRef(),
+            &p2p);
+        if (!(p2p && Queue->get_device().get_platform().get_backend() ==
```

Comment:
As it is currently implemented piextP2P does not take account of whether the two devices share the same backend.  Currently it is not supported for P2P memory copy across devices not sharing the same backend (enforced the additional check on line 958).

Reply:
@romanovvlad has a point. If someone was to go in and implement `piextP2P` for one of the other backends that actually care about the devices being passed, it's not unreasonable to assume that both devices are valid devices from that backend and I don't think they'd have a good way of telling whether or not they are. I think it would make sense to do the backend check before calling `piextP2P`.

Reply:
Yes it makes sense.  I made the change.